Exclude marketing channels in "Installed extensions" and "Discover more marketing tools" cards

ecgan opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/34907.

In this PR, we filter out marketing channels in the "Installed extensions" and "Discover more marketing tools" cards. This is because marketing channels already show up in its own "Channels" card.

As a result, there should be no more "Sales channels" tab in the "Discover more marketing tools" card. Each extension should only appear in one card.

Screenshot:

![image](https://user-images.githubusercontent.com/417342/223767871-5a2133e3-ac2c-4178-944b-4955176faa9c.png)

### How to test the changes in this Pull Request:

<!-- Please include detailed instructions on how these changes can be tested, make sure to review and follow the guide for writing high-quality testing instructions below. -->

- [x] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

Pre-condition: 

- Enable WooCommerce Multichannel Marketing experience in WooCommerce Settings > Advanced > Features (`/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`).

Test instructions:

1. Go to Marketing page at `/wp-admin/admin.php?page=wc-admin&path=%2Fmarketing`.
2. Each extension should only appear in one card.
3. There should be no "Sales channel" tab in the "Discover more marketing tools" card.

<!-- End testing instructions -->

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [x] Have you included testing instructions?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

ecgan committed:
Filter installed plugins against marketing channels in useInstalledPlugins.

ecgan committed:
Filter out marketing channels in useRecommendedPlugins.

ecgan committed:
Add changelog.

ecgan commented:
I have rebased the commits onto trunk branch, and changed this PR's base branch from `feature/34904-marketing-introduction-banner` to `trunk`, so that this PR does not depend on other PR and is ready for review.

eason9487 commented on plugins/woocommerce-admin/client/marketing/hooks/useInstalledPlugins.ts:
```diff
@@ -18,22 +21,36 @@ export type UseInstalledPlugins = {
 /**
  * Hook to return plugins and methods for "Installed extensions" card.
+ *
+ * The list of installed plugins will not include registered and recommended marketing channels.
```
From the perspective of this hook, it's named `useInstalledPlugins` but filters out some plugins internally, which might surprise the hook users.

While this can be clarified by this code comment, it might be possible to rename this hook, or to add hook parameters to indicate whether filtering out the registered and/or recommended ones.

eason9487 replied:
Another idea is to keep exposing the original `installedPlugins` and also exposing `installedPluginsWithoutChannels` separately.

ecgan replied:
> From the perspective of this hook, it's named `useInstalledPlugins` but filters out some plugins internally, which might surprise the hook users.
> 
> While this can be clarified by this code comment, it might be possible to rename this hook, or to add hook parameters to indicate whether filtering out the registered and/or recommended ones.

You got a point. I called it `useInstalledPlugins` because we uses the classic marketing wp.data store "wc/marketing" and it uses the term "plugin". I'm thinking I can rename them to `useInstalledExtensions` instead, because it is meant to be used with the "Installed extensions" card. Similarly, I'm thinking to rename `useRecommendedPlugins` to `useDiscoverTools`.

> Another idea is to keep exposing the original **`installedPlugins`** and also exposing `installedPluginsWithoutChannels` separately.

I less prefer this idea because we won't be using the `installedPlugins`.

eason9487 replied:
> I'm thinking I can rename them to `useInstalledExtensions` instead, because it is meant to be used with the "Installed extensions" card.

Renaming to `useInstalledExtensions` perhaps makes it more concrete in association with the "Installed extensions" card but it seems to still have the same confusion in the `useInstalledPlugins` naming: The `const { installedPlugins } = useInstalledExtensions()` statement still doesn't look like it will filter out some plugins.

eason9487 commented on plugins/woocommerce-admin/client/marketing/hooks/useInstalledPlugins.ts:
```diff
@@ -18,22 +21,36 @@ export type UseInstalledPlugins = {
+			dataRecommendedChannels,
+			( a, b ) => a.slug === b.product
+		)
+		.value();
```
Since `installedPlugins` are hydrated directly into the wp-data store via PHP preload data, the results of `useRegisteredChannels` and `useRecommendedChannels` will definitely be later than it. This causes the `installedPluginsWithoutChannels` to have incorrect data for a moment to a few seconds.

Although there is currently an additional loading state check in the location where this hook is used, which won't cause other problems in the runtime results, it's still recommended to fix it.

Demo a potential problem of flickering UI when data is not yet prepared:

https://user-images.githubusercontent.com/17420811/226311807-746c8159-4f39-4f85-a7c9-4f71bc5b4418.mp4

<br />

Demo code:

```ts
import '~/marketing/data';
import '~/marketing/data-multichannel';
import { InstalledExtensions } from './InstalledExtensions';
import './MarketingOverviewMultichannel.scss';

export const MarketingOverviewMultichannel: React.FC = () => {
	return (
		<div className="woocommerce-marketing-overview-multichannel">
			<InstalledExtensions />
		</div>
	);
};
```

eason9487 commented on plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/useRecommendedPlugins.ts:
This hook seems to have the same two concerns/issues as the `useInstalledPlugins`.

To make it easier to simulate/reproduce the problem of incorrect data, steps I tried:

```js
import { __unstableAwaitPromise } from '@wordpress/data-controls';

export function* getRecommendedChannels() {
	// Delay the resolver.
	yield __unstableAwaitPromise( new Promise( ( resolve ) => setTimeout( resolve, 3000 ) ) );
```

```ts
import { useRecommendedPlugins } from './DiscoverTools/useRecommendedPlugins';

export const MarketingOverviewMultichannel: React.FC = () => {
	const { plugins } = useRecommendedPlugins();
	console.log( 'plugins', plugins );
```


![image](https://user-images.githubusercontent.com/17420811/226321428-bb85f224-b130-4d9f-a8ce-54ecba68dd7d.png)

eason9487 reviewed (commented):
Thanks for the adjustment.

My main concerns are about two hooks may return incorrect data at certain times. Although the current runtime results on UI are shown correctly, it is still recommended to fix it first.

The two remaining ones about naming or usability are not blockers but are also recommended to adjust.

ecgan committed:
Fix incorrect data while loading in useInstalledPlugins.

ecgan committed:
Use fallback value instead of default value in useInstalledPlugins.

ecgan committed:
Rename UseInstalledPlugins to UseInstalledExtensions.

Also change related names from "plugin" to "extension".

ecgan committed:
Fix incorrect data while loading in useRecommendedPlugins.

See the incorrect data issue in https://github.com/woocommerce/woocommerce/pull/37126#discussion_r1141936304.

ecgan committed:
Change namings and add docs in useRecommendedPlugins.

ecgan committed:
Rename file useRecommendedPlugins to useRecommendedPluginsWithoutChannels.

ecgan committed:
Fix incorrect data while loading in useRecommendedPlugins.

See the incorrect data issue in https://github.com/woocommerce/woocommerce/pull/37126#discussion_r1141936304.

ecgan committed:
Change namings in useInstalledExtensions to useInstalledPluginsWithoutChannels.

ecgan committed:
Rename file UseInstalledPlugins to UseInstalledPluginsWithoutChannels.

ecgan committed:
Merge branch 'trunk' into feature/34907-marketing-exclude-channels

Conflicts:
	plugins/woocommerce-admin/client/marketing/hooks/index.ts

ecgan commented:
@eason9487 , thanks for the sharp eyes! And the demo code, video and screenshot, really appreciate those! 🙂 🙏 

I made quite some commits (some back and forth; naming is hard 😂 ) to address the problems. Here's a summary of the changes:

- `useInstalledPlugins` hook:
    - I renamed `useInstalledPlugins` hook to `useInstalledPluginsWithoutChannels`. Although the hook name is long, I think this shows the clearest meaning.
    - The `installedPlugins` property in the returned object is renamed to `data`. I change this to make it consistent with other hooks.
    - When loading is in progress, `data` will be an empty array. It returns the final correct data when loading is completed. So there should be no intermediate incorrect data.

- `useRecommendedPlugins` hook:
    - I renamed `useRecommendedPlugins` to `useRecommendedPluginsWithoutChannels`.
    - The `plugins` property in the returned object is renamed to `data`.
    - When loading is in progress, `data` will be an empty array. It returns the final correct data when loading is completed. So there should be no intermediate incorrect data.

This is now ready for another round of review. 🙏

eason9487 commented on plugins/woocommerce-admin/client/marketing/hooks/UseInstalledPluginsWithoutChannels.ts:
This filename was missed to start with a lowercase.

ecgan replied:
This is fixed, the result is in [`7637977` (#37126)](https://github.com/woocommerce/woocommerce/pull/37126/commits/76379770a184a307bb091b9b9d9eef952a365888).

eason9487 commented on plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/DiscoverTools.test.tsx:
```diff
@@ -20,37 +20,37 @@ jest.mock( '@woocommerce/components', () => {
-		( useInstalledPlugins as jest.Mock ).mockReturnValue( {
-			loadInstalledPluginsAfterActivation: jest.fn(),
+		( useInstalledPluginsWithoutChannels as jest.Mock ).mockReturnValue( {
+			loadInstalledExtensionsAfterActivation: jest.fn(),
```
This test suite mocks four non-existent functions `loadInstalledExtensionsAfterActivation` and it can still pass the tests. Looks like it doesn't need to mock the return value of function in each test?

Simply returning an empty object on the top of the test should be ok:

```js
jest.mock( '~/marketing/hooks', () => ( {
	useInstalledPluginsWithoutChannels: jest.fn( () => ( {} ) ),
} ) );
```

ecgan replied:
Thanks! Fixed in [`b18e170` (#37126)](https://github.com/woocommerce/woocommerce/pull/37126/commits/b18e1705cd9c32274a1ab6ed5af70010ee04f381).

eason9487 commented on plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/PluginsTabPanel.tsx:
```diff
@@ -60,7 +60,10 @@ export const PluginsTabPanel = ( {
-	const { loadInstalledPluginsAfterActivation } = useInstalledPlugins();
+	const {
+		loadInstalledPluginsAfterActivation:
+			loadInstalledExtensionsAfterActivation,
```
Is there a particular reason to rename this function or was is missed to further rename to the latest name?

ecgan replied:
Thanks, missed this when I doing the renaming. Fixed in [`5e6cd11` (#37126)](https://github.com/woocommerce/woocommerce/pull/37126/commits/5e6cd11028fe6219c264de067b1550c6f097903a).

eason9487 commented on plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/useRecommendedPluginsWithoutChannels.ts:
```diff
@@ -0,0 +1,91 @@
+		}, [] );
+
+		const isInitializing =
+			( loadingRecommendedPlugins && ! dataRecommendedPlugins.length ) ||
```
If API will never return an empty array as `dataRecommendedPlugins`, this logic should be fine, and it would be a bit nicer to note this here.

But if any case the `dataRecommendedPlugins` was an empty array, the `isInitializing` would not fit with its description:

https://github.com/woocommerce/woocommerce/blob/e281c09bc67eb3b94fde3bf82a93f358ea2b2ae3/plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/useRecommendedPluginsWithoutChannels.ts#L15-L18

ecgan replied:
Thanks. The API data may be an empty array. I have changed the description in [`0149252` (#37126)](https://github.com/woocommerce/woocommerce/pull/37126/commits/01492520d326232c98806ed37ae90b6b7b95c73e).

eason9487 reviewed (commented):
Thanks for the update.

Most of the code looks good. The only needed tweak is the filename `UseInstalledPluginsWithoutChannels.ts` doesn't follow the convention of React hook. The rest comments are minor things or a question.

I guess this repo is ticked the "Dismiss stale pull request approvals when new commits are pushed" option for the `trunk` branch, so even though I would like to approve it in this round but probably it would still need a new approval after pushing the file renaming commit. So feel free to ping me afterward.

ecgan committed:
Fix file name for useInstalledPluginsWithoutChannels.

ecgan committed:
Fix and simplify mocks in DiscoverTools tests.

ecgan committed:
Simplify naming in PluginsTabPanel.

ecgan committed:
Fix JSDOC comment in UseRecommendedPluginsWithoutChannels.

ecgan committed:
Reorganize code.

ecgan committed:
Temporary rename file to fix casing issue.

ecgan committed:
Fix hook file name with correct casing.

ecgan commented:
@eason9487 , thanks, this is ready for another round. 🙂 🙏

eason9487 reviewed (approved):
LGTM.
